src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -335,35 +340,54 @@ void KafkaSink::consume(Chunk chunk)
             continue;
         }
 
-        /// Compute and collect message keys and removed the key column from the block after executing the expression
+        /// Compute and collect message keys.
         message_key_expr->execute(block_with_shard.block);
-        auto message_key_column {block_with_shard.block.getByPosition(message_key_column_pos).column};
-        key_columns.push_back(message_key_column);
-
-        // Collect all the message keys for creating the messages for the block.
+        auto message_key_column {block_with_shard.block.getByName(message_key_column_name).column};
         size_t rows {message_key_column->size()};
         for (size_t i = 0; i < rows; ++i)
             keys_for_current_batch.push_back(message_key_column->getDataAt(i));
 
-        if (delete_message_key_column)
-            block_with_shard.block.erase(message_key_column_pos);
+        /// After `message_key_expr->execute`, the columns in `block_with_shard.block` could be out-of-order.
+        /// We have to make sure the the column order in `block_with_shard.block` exactly matches the order in header,
+        /// otherwise, the output format writer will panic.
+        Block blk;
+        blk.reserve(getHeader().columns());
+        for (const auto & col : getHeader())
+            blk.insert(std::move(block_with_shard.block.getByName(col.name)));
 
-        writer->write(block_with_shard.block);
+        writer->write(blk);
     }
 
     /// With `wb->setAutoFlush()`, it makes sure that all messages are generated for the chunk at this point.
-    auto n = rd_kafka_produce_batch(
+    rd_kafka_produce_batch(
         topic.get(),
         RD_KAFKA_PARTITION_UA,
-        RD_KAFKA_MSG_F_PARTITION | RD_KAFKA_MSG_F_BLOCK,
+        RD_KAFKA_MSG_F_FREE | RD_KAFKA_MSG_F_PARTITION | RD_KAFKA_MSG_F_BLOCK,
         current_batch.data(),
         current_batch.size());
 
-    if (static_cast<size_t>(n) != current_batch.size())
-        /// Find the first error and throw it.
-        for (const auto & msg : current_batch)
-            if (msg.err)
-                throw Exception(klog::mapErrorCode(msg.err), rd_kafka_err2str(msg.err));
+    rd_kafka_resp_err_t err {RD_KAFKA_RESP_ERR_NO_ERROR};
+    for (size_t i = 0; i < current_batch.size(); ++i)
+        if (current_batch[i].err)
+            err = current_batch[i].err;
+        else
+            batch_payload[i].release(); /// payload of messages which are succesfully handled by rd_kafka_produce_batch will be free'ed by librdkafka
+
+    /// Clean up all the bookkeepings for the batch.
+    std::vector<rd_kafka_message_t> batch;
```

Comment:
Since we do clean up here, do we need do again at line 309

Reply:
Just in case the function call gets interrupted somehow and cause the cleanup skipped ( not 100% sure if that will happen or not ). As discussed, keep this as it is for now.